src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file VST3Effect.cpp
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+#include "VST3Effect.h"
+
+#include <stdexcept>
+#include <wx/log.h>
+#include <pluginterfaces/vst/ivstaudioprocessor.h>
+#include <pluginterfaces/vst/ivsteditcontroller.h>
+#include <pluginterfaces/vst/ivstprocesscontext.h>
+#include <wx/msw/spinctrl.h>
+
+#include "internal/ComponentHandler.h"
+#include "internal/ParameterChanges.h"
+#include "internal/PlugFrame.h"
+
+#include "ShuttleGui.h"
+
+#include "VST3EffectsModule.h"
+#include "VST3Utils.h"
+
+void VST3Effect::Initialize()
+{
+   using namespace Steinberg;
+
+   const auto& pluginFactory = mModule->getFactory();
+   if(auto effectComponent = pluginFactory.createInstance<Vst::IComponent>(mEffectClassInfo.ID()))
+   {
+      if(effectComponent->initialize(VST3EffectsModule::LocalContext()) == kResultOk)
+      {
+         mEffectComponent = effectComponent;
+
+         Vst::IEditController* editController;
+         if(effectComponent->queryInterface(Vst::IEditController::iid, reinterpret_cast<void**>(&editController)) != kResultOk)
+         {
+            //Plugin does not have a separate edit controller
+            TUID controllerCID;
+            
+			   if (effectComponent->getControllerClassId (controllerCID) == kResultTrue)
+			   {
+				   if(mEditController = pluginFactory.createInstance<Vst::IEditController> (VST3::UID (controllerCID)))
+				   {
+                  if(mEditController->initialize(VST3EffectsModule::LocalContext()) != kResultOk)
+                  {
+                     wxLogMessage("Cannot create edit controller: initialization failed");
+                     mEditController = nullptr;
+                  }
+				   }
+               else
+                  wxLogMessage("Cannot create edit controller: attempt failed");
+			   }
+            else
+               wxLogMessage("Cannot create edit controller: failed read controller Class ID from the effect component");
+         }
+         else
+            mEditController = editController;
+      }
+      else
+         throw std::runtime_error("Cannot initialize VST3 effect component");
+   }
+   else
+      throw std::runtime_error("Cannot create VST3 effect component");
+
+   if(mEditController)
+   {
+      mComponentHandler = owned(safenew internal::ComponentHandler);
+      mEditController->setComponentHandler(mComponentHandler);
+   }
+
+   //defaults
+   mSetup.processMode = Vst::kOffline;
+   mSetup.symbolicSampleSize = Vst::kSample32;
+   mSetup.maxSamplesPerBlock = 512;
+   mSetup.sampleRate = 44100.0;
```

Comment:
I see how three of these values are ultimate communicated to the VST3 library in VST3ProcessBlock, but I don't see where the sample rate is given to it.  Is that an omission?

Have you tested with non-default sample playback rate?  (Change it in the drop-down at bottom left)

If you have an effect like an equalizer, but VST3 is misinterpeting the sample rate, then you might hear wrong results at non-default sample rate choices.

Reply:
I believe that `VST3Effect::SetSampleRate` and `VST3Effect::SetBlockSize` are used to set up the effect. Yes, I've tested effects with non-default sample rates.